Shared lib cache

tbitonti opened the pull request:
Re-use shared library containers between applications.  Containers used by shared libraries are tracked by path using reference counts.  Each use of a shared library increases the reference count.  Reference counts are reduced when uninstalling an application.  (Or, when the application is garbage collected, if the usual install/uninstall APIs fail.)

FATs are coded and included in the PR.  See updates, below, for a test summary.

tbitonti committed:
Deferred cache prototype.

tbitonti committed:
Add API to count library container references and to release these when
application information goes out of use.

tbitonti commented:
#build

tbitonti committed:
Test shared library files for changes;
Initial ShalredLibTest FAT test.

tbitonti committed:
Complete shared LIB fat:
* test appliction start/stop with shared library

tbitonti committed:
Add shared library variations tests.
Verify access to shared libraries.

tbitonti committed:
Enable trace and verify capture and release activity.

tbitonti commented:
#build

tbitonti commented:
#build

tbitonti committed:
Correctly obtain counts for logging within the cache lock region.

tbitonti commented:
<h3>FAT Testing</h3>

FAT tests are performed by the new application manager FAT class:

* open-liberty/dev/com.ibm.ws.app.manager_fat/fat/src/componenttest/application/manager/test/SharedLibTest.java
  * testSharedLibs()

For the test, the server is initially configured with sixteen applications and three shared libraries, and with four library JARs.

The state of the shared library cache is examined immediately to verify that the shared library cache has the expected state to verify that the expected initial operations were performed.

The initial verification shows that the cache operates correctly for a usual, static, configuration.

Next, using the same server, the configuration is updated to change details of the shared library configuration.  The test cycles several times through four configurations:

* one application and one shared library
* four applications and two shared libraries
* eight applications and two shared libraries
* sixteen applications and three shared libraries

The tests use four different shared library elements.

The cache state and of cache operations are verified after each update.

tbitonti committed:
Compute reference counts correctly when out-of-order trace is written by the cache.

tbitonti commented:
Build was good, except for a problem handling out-of-order cache trace output.  Correcting and rerunning.

tbitonti commented:
#build

tbitonti committed:
Add Trivial where needed.

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/CollectionUtils.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.AbstractMap;
+import java.util.Map;
+
+//@formatter:off
```
Why does this solution circumvent the eclipse project's code format configuration rather than modify it?

tbitonti replied:
CollectionUtils is no longer in use and has been removed.
(Generally, the eclipse formatting often badly formats code; for entirely new classes, turning off the auto-format is easier than fighting it.)

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/CaptureCache.java:
```diff
@@ -0,0 +1,434 @@
+ */
+public class CaptureCache<T> {
+    // Trace with the application manager.
+    // TODO: Move DeferredCache to a more common location.
```
Remove this comment. DeferedCache.java is not a part in this solution.   Also, the previous comment about Trace is superfluous as it repeats exactly what the code is doing.

tbitonti replied:
Removed comment.

tbitonti committed:
Add JAR update test for shared lib caching.

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/CaptureCache.java:
```diff
@@ -0,0 +1,434 @@
+        }
+    }
+
+    // Store a caching Supplier.  That moves production outside of management of
```
Perhaps I misunderstood this comment. It seems more appropriate to the class's javadoc description as it explains why the cache utilizes Supplier semantics.  Also, is the TODO still a concern?  It's not a bad idea to track library PID as it will enhance serviceability, especially when examining the cache a heap dump.  If you can correlate the libPath for a cache entry using  the trace and/or the introspection, then I don't see a reason for concern.  Trace seemed to have all the necessary lib info when a cache entry is acquired.

tbitonti replied:
The java doc for 'storage' was clarified.

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/CaptureCache.java:
```diff
@@ -0,0 +1,434 @@
+     * Key data and supplier storage have linked state and are
+     * satisfied by a single lock.
+     */
+    private final Object cacheLock;
```
The CaptureCache instance is shared among the deployed apps in the app manager.  If there's contention on this lock, consider revising the solution to use a read/write lock or a more granular lock that manages cache access using  libPath.

tbitonti replied:
For now, a simple object lock seems sufficient.  Requests for containers for shared libraries are not expected to be performed with a high enough frequency for the lock implementation to make a difference.

tbitonti commented:
#build

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/CaptureCache.java:
```diff
@@ -0,0 +1,434 @@
+ *
+ * Callers must reliable release suppliers to avoid memory leaks.
+ */
+public class CaptureCache<T> {
```
FWICT, the CaptureCache class is used to pool container suppliers keyed by library path and to maintain a reference count for each supplier. It's an object pool of sort, and I think the solution in DeployedAppInfoBase should mention that or rename the instance variable to reflect its usage.

tbitonti replied:
Ah, it's not a pool.  Each of the containers is actively in use, and is discarded as soon as the use count reaches zero.  None of the producers or containers is retained in a pool for later reuse.

dazavala replied:
Please pardon the error.  Definitely not a pool.

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/CaptureCache.java:
```diff
@@ -0,0 +1,434 @@
+            } else {
+                uniqueKey = keyData.getKey();
+                newCount = keyData.decrement();//++
+                if ( newCount == 0 ) {
```
What happens if count goes negative?  I didn't see any code that would prevent this unexpected condition. 
 Consider changing to `if (newCount <= 0) {`, prohibit a negative count, and emit a debug warning if the condition occurs.

tbitonti replied:
Additional logic added to handle a '< 0' case.

dazavala replied:
Thanks.  I've seen many build breaks involving fields that contain inexplicable values.

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/DeployedAppInfoBase.java:
```diff
@@ -118,15 +122,137 @@ public List<ContainerInfo> getClassesContainerInfo() {
+     * Use of a static cache requires that all library containers
+     * for a given file path are identical.
+     */
+    protected static final CaptureCache<LibraryContainerData> containerCache = new CaptureCache<>("container");
```
Names of operations and fields become more general here.  This is a libraryContainerPool or libraryContainerCache.  I make similar comments below.

tbitonti replied:
Renamed from 'containerCache' to 'libraryContainerCache'.

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/DeployedAppInfoBase.java:
```diff
@@ -118,15 +122,137 @@ public List<ContainerInfo> getClassesContainerInfo() {
+     * @return A new supplier of the library container which will capture the
+     *         result of the base supplier.
+     */
+    protected static CaptureSupplier<LibraryContainerData> capture(String libPath,
```
Name should indicate function is specific to library support, like captureLibrary or captureLibraryContainer.  Same for the relase methods.

tbitonti replied:
Rename to 'captureLibraryContainer' and 'releaseLibraryContainers'.

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/DeployedAppInfoBase.java:
```diff
@@ -118,15 +122,137 @@ public List<ContainerInfo> getClassesContainerInfo() {
 
+    // Data for caching a container ...
+
+    protected static class LibraryContainerData {
```
Don't need the protected access here.  Also, any reason why the class does not declare private access to its fields along with public accessor methods?

tbitonti replied:
Protected in case a subclass has a need to use this data structure.
'public final' with no accessors to keep the class lightweight.  It's intended to be a lightweight record type data structure.

dazavala commented on dev/com.ibm.ws.app.manager.module/src/com/ibm/ws/app/manager/module/internal/DeployedAppInfoBase.java:
```diff
@@ -175,88 +299,238 @@ public List<ContainerInfo> getClassesContainerInfo() {
+        // containers will have data specific to the application, and likely
+        // cannot be shared.
+
+        private Container setupContainer(String libraryPid, File libraryFile) {
```
An observation: It's not clear to me whether this revised implementation of setupContainer() returns null for each of the conditions it used to.  That could be a change in behavior for calling method, `getContainers()`.

tbitonti replied:
The cases *should* be the same; the intent was to answer null in exactly the same cases as before.  The main change is the addition of trace to the base container creation steps, which were added as an assist. That the container can ever be null is problematic.  That means we are simply ignoring class path entries that fail to resolve to a container for whatever reason.  The added trace is very useful when diagnosing problems in the container resolution.

tbitonti replied:
Even in the case of dynamic updates, the new code should still obtain new contains every time the same container is resolved and the underlying file was updated (that is, if the file size or time stamp was updated).  Each call to resolve the container will check for changes and will only answer the cached container if there are no changes.

dazavala reviewed (changes requested):
Hey, @tbitonti.  I'll approve the solution once you've had a look at my comments.  I traced various scenarios we discussed and some we did not.  AFAIS the solution shares library container instances for common library paths among all applications that reference the libraries, and it correctly maintains reference counts to the containers.  Though the FAT verifies reference counting, I observed the solution maintains reference counts as expected by managing apps with the admin console.  Traces look serviceable; I did not dump the server to see the introspection.

I think you should strongly consider safeguarding the cached supplier (container) reference count from going negative.  Where possible remove or reduce the comments.  The code is readable.  Also, as I do not typically service the app manager I would like someone more familiar with the App Mgr to offer a second opinion.

Some concerns we've discussed:
1. What happens when a library path is shared by multiple apps via common lib reference, and the jar file deployed in that path is updated?  The app manager appears insensitive to dynamic updates to library jars.  I don't see any code that would force an app restart in this scenario.  And if the app manager were sensitive, then it must restart all apps that reference the library, just as it does when a `<library/>` config is changed dynamically.  

2. What about apps that reference private libraries or the global shared library?  I looked into this.  The app manager processes containers for private libraries and global libraries the same way it does for common libraries.  Running a few scenarios I saw no difference in how solution acquires and releases (pools) library containers for any of these library types, even when dynamically changing lib references from one type to another, including the global lib.

> FYI: It would be easy to add global and private library refs to the test scenarios.  Apps need not reference the `global` library.  Deploy a jar to `${server.config.dir}/lib/global`.  The jar will be visible to all deployed apps lacking a  `<classloading/>` element.  Apps that have a `<classloading/>` config must reference the `global` library to see the jar, and the deployment must declare a `<library/>` with `id=global`.  

3. What happens when a library path is updated in the library config?  WRT to app mgmt, any app referencing the library is restarted. For common lib references the restarted app classloaders will delegate to a common lib classloader having the new lib path (container) in its classpath; for private libs, the loaders will have the new lib path on their classpaths; and for global library, the loaders will share a parent "gsl" loader  having the new lib path in its classpath.
 
4. Any cache (pool) is vulnerable to leakage and inconsistency. What's happens if the user replaces a jar in an existing lib path and then restarts an app that references the lib, where the lib is also shared by another app.  Does the app manager pick up the new jar content -- i.e. does it send the updated the container to the new app classloader -- and will this cause an inconsistency with the cached container supplier?

tbitonti committed:
Updates per review comments.
Comment updates.
Changed 'container' to 'libraryContainer' for a variable and several methods.
Added logic and warning for unexpected reference counts being negative.

dazavala reviewed (approved):
Approved.  All concerns addressed.

tbitonti commented:
#build

tbitonti commented:
#libby

tbitonti commented:
#build

tbitonti commented:
Build has a single unrelated failure:

```
no component | 97336 | 0 | 1 |   |  
com.ibm.ws.springboot.fat20_fat FAT tests | 1 | 0 | 1 |   |  
com.ibm.ws.springboot.support.fat.FATSuite | 1 | 0 | 1 | 0 s | 1
testBlockingIOServlet40

```

tbitonti commented:
Fixes #PH55953
